src/core/net/tcp6.hpp

```diff
@@ -311,6 +310,36 @@ class Tcp : public InstanceLocator, private NonCopyable
          * @retval kErrorFailed  Failed to deinitialize the TCP endpoint.
          */
         Error Deinitialize(void);
+
+    private:
+        enum : uint8_t
+        {
+            kNumTimers = 4,
+        };
+
+        static_assert(sizeof(mTimers[0]) >= sizeof(TimerMilli), "mTimers entries are too small");
+        static_assert(OT_ARRAY_LENGTH(mTimers) == kNumTimers, "mTimers has an incorrect array length");
+
+        TimerMilli &GetTimer(size_t aIndex) { return *reinterpret_cast<TimerMilli *>(&mTimers[aIndex].mSize[0]); }
+
+        static Endpoint &FromTimer(TimerMilli &aTimer, size_t aIndex)
+        {
+            uint8_t *ptr = reinterpret_cast<uint8_t *>(&aTimer);
+            ptr -= offsetof(Endpoint, mTimers[aIndex].mSize[0]);
+            return *reinterpret_cast<Endpoint *>(ptr);
+        }
+
+        TimerMilli &     GetDelackTimer() { return GetTimer(0); }
+        static Endpoint &FromDelackTimer(TimerMilli &aTimer) { return FromTimer(aTimer, 0); }
+
+        TimerMilli &     GetRexmtPersistTimer() { return GetTimer(1); }
+        static Endpoint &FromRexmtPersistTimer(TimerMilli &aTimer) { return FromTimer(aTimer, 1); }
+
+        TimerMilli &     GetKeepTimer() { return GetTimer(2); }
+        static Endpoint &FromKeepTimer(TimerMilli &aTimer) { return FromTimer(aTimer, 2); }
+
+        TimerMilli &     Get2MslTimer() { return GetTimer(3); }
```

Comment:
I like this approach, and I think it is the one we should go for, for now. See also my more detailed response to your previous comment.

Reply:
> My understanding is that, on expiry, it would iterate over all TCP endpoints and check the four timers on each one to check which ones expired. And when a timer expires or is cancelled, it would iterate over all TCP endpoints to find the timer that expires next. 

My suggestion is share one timer per end-point (instead of 4 timers per end point).

Reply:
I see. If we do that, we'd still have to find a way to put one `Timer` object in each TCP endpoint though.

What are your thoughts on using a single `Timer` for all TCP endpoints? The advantage there is that the timer would not be in the `otTcpEndpoint` struct. This would eliminate the problems you described earlier and the drawbacks of the solution that I found.

Reply:
I was not considering one `Timer` for all TCP end point but I agree that can also work. :)

As you said the challenge is its scalability. I would guess if we have smaller number of TCP end points, (say in range 10s-20s end points) then maybe iterating over all would be fine.

Reply:
Regarding scalability, the current `Timer` implementation also has to deal with this to some extent. Although it uses a sorted linked list, I believe it does require iterating over the linked list to add new timers or cancel timers. We can also keep the existing Endpoint linked list sorted in order of each endpoint's first expiring timer.

Reply:
Exactly! `Timer::Scheduler` keeps track of sorted linked list of all scheduled `Timer` objects and may need to iterate through the list from some methods. We don't want to have too many timers (this is basically limited by the number of modules in OT).
- As example, we added `TimeTicker` (from https://github.com/openthread/openthread/pull/5379) which helped us remove a bunch of timers and share one among a bunch of modules.

Reply:
OK. Thanks for the context that OpenThread tries to avoid having too many `Timer`s.

Reply:
Agree with @abtink . Actually, #6036 uses one Timer for all TCP sockets.

Reply:
The latest update to the code below also uses one Timer for all TCP sockets. Thanks again for the guidance here.